packages/jest-fake-timers/src/jestFakeTimers.ts

```diff
@@ -432,20 +427,17 @@ export default class FakeTimers<TimerRef> {
       return null;
     }
 
-    const uuid = this._uuidCounter++;
+    const uuid = String(this._uuidCounter++);
 
     this._immediates.push({
       callback: () => callback.apply(null, args),
-      uuid: String(uuid),
+      uuid,
     });
 
-    const cancelledImmediates = this._cancelledImmediates;
+    const immediates = this._immediates;
     this._timerAPIs.setImmediate(() => {
-      if (!cancelledImmediates.hasOwnProperty(uuid)) {
-        // Callback may throw, so update the map prior calling.
-        cancelledImmediates[String(uuid)] = true;
-        callback.apply(null, args);
-      }
+      this._fakeClearImmediate(uuid);
+      immediates.find(x => x.uuid === uuid) && callback.apply(null, args);
```

Comment:
Can we make this an if block around the cancellation and call again? I think this is an example of code that looks smart but is fragile, e.g. if `_fakeClearImmediate` used `splice` instead of changing the `_immediates` reference, it would not find the uuid here.